README.md
**Note: This library has been superseded by [blp](https://github.com/matthewgilbert/blp), the next iteration of pdblp.**. This library is no longer under active development.

pdblp
=====
[![PyPI version](https://badge.fury.io/py/pdblp.svg)](https://badge.fury.io/py/pdblp)

This is a simple interface to integrate pandas and the Bloomberg Open API.
The standard Bloomberg API provides an extensive set of features for building
applications on top of however does not provide easy and interactive access to
data. This package provides several functions for accessing historical market
data and reference data. A simple set of examples is available
[here](https://matthewgilbert.github.io/pdblp/tutorial.html).

The library borrows heavily from a similar package available
[here](https://github.com/kyuni22/pybbg)

## Questions

`pdblp` simply wraps `blpapi` responses into `pandas` dataframes, general
questions about `blpapi` **should not** be posted as issues. The best way to get
support for `blpapi` is either from Bloomberg directly via the terminal or on
Stack Overflow, using tag `blpapi` (and `pdblp` if it also relates to the `pdblp` 
library).

## Requires

`python 3.x`

[Bloomberg Open API](http://www.bloomberglabs.com/api/)

[pandas](http://pandas.pydata.org/)

and for `pdblp.parser`

[pyparsing](https://pythonhosted.org/pyparsing/) >= 2.2.0

## Installation
You can install from PyPi using

```
pip install pdblp
```

or from conda-forge using

```
conda install -c conda-forge pdblp
```

or you can clone this repository and pip install the package, i.e.

```
git clone https://github.com/matthewgilbert/pdblp.git
pip install -e pdblp
```

`blpapi` can be installed directly from the above link or you can install from
the `conda-forge` channel on Anaconda, e.g.

```
conda install -n blp_env -c conda-forge blpapi
```

If going this route make sure there is a build available for your version of
python https://anaconda.org/conda-forge/blpapi/files

If you are getting `'GLIBCXX_3.4.21' not found` error when attempting to import
`blpapi`, the following appears to resolve this issue.

```
conda install libgcc
```

## Documentation

The documentation can be viewed at https://matthewgilbert.github.io/pdblp/

### Bloomberg Documentation

For general documentation on the Bloomberg API check out the Developer's Guide.
For documentation on relevant Bloomberg fields for accessing data, check out
the Reference Guide: Services and Schemas. To access these, from a
Bloomberg Terminal go `WAPI <GO>` -> `API Developer's Guide`.

### Building the documentation

The documentation relies on [Sphinx](http://www.sphinx-doc.org/en/master/).
Building the documentation can be done by setting up a conda environment using
`conda create --name pdblp_doc --file doc-environment.yml`, sourcing this
environment and then installing the relevant version of `pdblp`. Documentation
can then be built using

```
cd doc
make html
```

and viewed in ./doc/_build. Before building this ensure that you are
logged into a Bloomberg terminal as this is required for building many of the
examples.

## Citing `pdblp`

If you make use of `pdblp` in publications, citations to this package would be
appreciated. You can use the badge below to generate a DOI and reference text
for the latest related version of pdblp.

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2529925.svg)](https://doi.org/10.5281/zenodo.2529925)


CHANGELOG
# pdblp 0.1.1

- fixed bug where list of parameters persisted over multiple calls

# pdblp 0.1.2

- added context management support with bopen()
- added bsrch() for Bloomberg SRCH Data
- added parsing to JSON of string representation of Request and Response
messages with parser.py module
- modified ref() to return NaN instead of throwing an error when
fieldException of NOT_APPLICABLE_TO_REF_DATA is given
- modified ref_hist() to support any historical reference date override
instead of just REFERENCE_DATE, e.g. CURVE_DATE

# pdblp 0.1.3

- expose blpapi.Event.TIMEOUT to user through timeout parameter in BCon()
instantiation

# pdblp 0.1.4

- Fix bug accessing null blpapi.element.Element in ref(), see issue #32
- Raise a ValueError for bad inputs to bdh(), see issue #33

# pdblp 0.1.5

- Seperate ref() into bulkref() and ref(), see issue #35
- Update tutorial docs to reflect new functions
- Update Sphinx build process to use sphinx 1.7.5 and provide environment
file for building docs

# pdblp 0.1.6

- Add ability to chain start() method, e.g. con = BCon().start()
- Fix logger to interface with upstream logging libraries
- Add pull request template
- Add NaN support to parser grammar
- Support empty bulkref() requests

# pdblp 0.1.7

- Automate conda and pypi deploys with TravisCI
- Change from unittest to using pytest
- Seperate offline tests which do not require BBG connection
- Refactor event handling and logging into _receive_events()
- Remove restart() method, see issue #46
- Allow constructor to take a blpapi.Session and blpapi.Identity, for SAPI

# pdblp 0.1.8

- Fix bopen() to work with passed session
- Refactor pdblp to use blpapi.Messages parsed into dicts
- Remove conda package building
- Remove testing for python 3.5 and 3.4

# pdblp 0.1.9

 - Allow arbitrary dates formats


pdblp/tests/conftest.py
def pytest_addoption(parser):
    parser.addoption(
        "--offline", action="store_true", default=False,
        help="Run tests in offline mode (has no BBG connection)"
    )
    parser.addoption(
        "--port", action="store", default=8194, type=int,
        help="Port to connect"
    )
    parser.addoption(
        "--host", action="store", default='localhost', help="Host to connect"
    )
    parser.addoption(
        "--timeout", action="store", default=5000, type=int,
        help="BCon.timeout value"
    )


def pytest_configure(config):
    config.cache.set('offline', config.getoption('--offline'))


pdblp/tests/test_pdblp.py
import pytest
import pandas as pd
import numpy as np
from pandas.util.testing import assert_frame_equal, assert_index_equal
from pdblp import pdblp
import blpapi
import os


@pytest.fixture(scope="module")
def port(request):
    return request.config.getoption("--port")


@pytest.fixture(scope="module")
def host(request):
    return request.config.getoption("--host")


@pytest.fixture(scope="module")
def timeout(request):
    return request.config.getoption("--timeout")


@pytest.fixture(scope="module")
def con(host, port, timeout):
    return pdblp.BCon(host=host, port=port, timeout=timeout).start()


@pytest.fixture(scope="module")
def data_path():
    return os.path.join(os.path.dirname(__file__), "data/")


def pivot_and_assert(df, df_exp, with_date=False):
    # as shown below, since the raw data returned from bbg is an array
    # with unknown ordering, there is no guruantee that the `position` will
    # always be the same so pivoting prior to comparison is necessary
    #
    # fieldData = {
    #     INDX_MWEIGHT[] = {
    #         INDX_MWEIGHT = {
    #             Member Ticker and Exchange Code = "BON8"
    #             Percentage Weight = 2.410000
    #         }
    #         INDX_MWEIGHT = {
    #             Member Ticker and Exchange Code = "C N8"
    #             Percentage Weight = 6.560000
    #         }
    #         INDX_MWEIGHT = {
    #             Member Ticker and Exchange Code = "CLN8"
    #             Percentage Weight = 7.620000
    #         }
    #     }
    # }
    name_cols = list(df_exp.name.unique())
    sort_cols = list(df_exp.name.unique())
    index_cols = ["name", "position", "field", "ticker"]
    if with_date:
        sort_cols.append("date")
        index_cols.append("date")

    df = (df.set_index(index_cols).loc[:, "value"]
          .unstack(level=0).reset_index().drop(columns="position")
          .sort_values(by=sort_cols, axis=0))
    df_exp = (df_exp.set_index(index_cols).loc[:, "value"]
              .unstack(level=0).reset_index().drop(columns="position")
              .sort_values(by=sort_cols, axis=0))
    # deal with mixed types resulting in str from csv read
    for name in name_cols:
        try:
            df_exp.loc[:, name] = df_exp.loc[:, name].astype(float)
        except ValueError:
            pass
    for name in name_cols:
        try:
            df.loc[:, name] = df.loc[:, name].astype(float)
        except ValueError:
            pass
    if with_date:
        df.loc[:, "date"] = pd.to_datetime(df.loc[:, "date"],
                                           format="%Y%m%d")
        df_exp.loc[:, "date"] = pd.to_datetime(df_exp.loc[:, "date"],
                                               format="%Y%m%d")
    assert_frame_equal(df, df_exp)


ifbbg = pytest.mark.skipif(pytest.config.cache.get('offline', False),
                           reason="No BBG connection, skipping tests")


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20180510", "2018-05-11"), ("2018/05/10", "11May2018"),
     (pd.datetime(2018, 5, 10), pd.Timestamp(2018, 5, 11))]
)
@ifbbg
def test_bdh_empty_data_only(con, start_date, end_date):
    df = con.bdh(
        tickers=['1437355D US Equity'],
        flds=['PX_LAST', 'VOLUME'],
        start_date=start_date,
        end_date=end_date,
        longdata=False
    )
    df_exp = pd.DataFrame(
        [],
        index=pd.DatetimeIndex([], name='date'),
        columns=pd.MultiIndex.from_product([[], []], names=('ticker', 'field'))
    )
    assert_frame_equal(df, df_exp)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20180510", "2018-05-11"), ("2018/05/10", "11May2018"),
     (pd.datetime(2018, 5, 10), pd.Timestamp(2018, 5, 11))]
)
@ifbbg
def test_bdh_empty_data_with_non_empty_data(con, start_date, end_date):
    df = con.bdh(
        tickers=['AAPL US Equity', '1437355D US Equity'],
        flds=['PX_LAST', 'VOLUME'],
        start_date=start_date,
        end_date=end_date,
        longdata=False
    )
    df_exp = pd.DataFrame(
        [[190.04, 27989289.0], [188.59, 26212221.0]],
        index=pd.DatetimeIndex(["20180510", "20180511"], name="date"),
        columns=pd.MultiIndex.from_product([["AAPL US Equity"],
                                            ["PX_LAST", "VOLUME"]],
                                           names=["ticker", "field"])
    )
    assert_frame_equal(df, df_exp)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20180215", "2018-02-16"), ("2018/02/15", "16Feb2018"),
     (pd.datetime(2018, 2, 15), pd.Timestamp(2018, 2, 16))]
)
@ifbbg
def test_bdh_partially_empty_data(con, start_date, end_date):
    df = con.bdh(
        tickers=['XIV US Equity', 'AAPL US Equity'],
        flds=['PX_LAST'],
        start_date=start_date,
        end_date=end_date,
        longdata=False
    )
    df_exp = pd.DataFrame(
        [[6.04, 172.99], [np.NaN, 172.43]],
        index=pd.DatetimeIndex(["20180215", "20180216"], name="date"),
        columns=pd.MultiIndex.from_product(
            [["XIV US Equity", "AAPL US Equity"], ["PX_LAST"]],
            names=["ticker", "field"]
        )
    )
    assert_frame_equal(df, df_exp)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20150629", "2015-06-30"), ("2015/06/29", "30Jun2015"),
     (pd.datetime(2015, 6, 29), pd.Timestamp(2015, 6, 30))]
)
@ifbbg
def test_bdh_one_ticker_one_field_pivoted(con, start_date, end_date):
    df = con.bdh('SPY US Equity', 'PX_LAST', start_date, end_date)
    midx = pd.MultiIndex(levels=[["SPY US Equity"], ["PX_LAST"]],
                         labels=[[0], [0]], names=["ticker", "field"])
    df_expect = pd.DataFrame(
        index=pd.date_range("2015-06-29", "2015-06-30"),
        columns=midx,
        data=[205.42, 205.85]
    )
    df_expect.index.names = ["date"]
    assert_frame_equal(df, df_expect)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20150629", "2015-06-30"), ("2015/06/29", "30Jun2015"),
     (pd.datetime(2015, 6, 29), pd.Timestamp(2015, 6, 30))]
)
@ifbbg
def test_bdh_one_ticker_one_field_longdata(con, start_date, end_date):
    df = con.bdh('SPY US Equity', 'PX_LAST', start_date, end_date,
                 longdata=True)
    idx = pd.Index(["date", "ticker", "field", "value"])
    data = [["2015-06-29", "2015-06-30"],
            ["SPY US Equity", "SPY US Equity"], ["PX_LAST", "PX_LAST"],
            [205.42, 205.85]]
    df_expect = pd.DataFrame(data=data, index=idx).transpose()
    df_expect.loc[:, "date"] = pd.to_datetime(df_expect.loc[:, "date"])
    df_expect.loc[:, "value"] = np.float64(df_expect.loc[:, "value"])
    assert_frame_equal(df, df_expect)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20150629", "2015-06-30"), ("2015/06/29", "30Jun2015"),
     (pd.datetime(2015, 6, 29), pd.Timestamp(2015, 6, 30))]
)
@ifbbg
def test_bdh_one_ticker_two_field_pivoted(con, start_date, end_date):
    cols = ['PX_LAST', 'VOLUME']
    df = con.bdh('SPY US Equity', cols, start_date, end_date)
    midx = pd.MultiIndex(
        levels=[["SPY US Equity"], cols],
        labels=[[0, 0], [0, 1]], names=["ticker", "field"]
    )
    df_expect = pd.DataFrame(
        index=pd.date_range("2015-06-29", "2015-06-30"),
        columns=midx,
        data=[[205.42, 202621332], [205.85, 182925106]]
    )
    df_expect = df_expect.astype(np.float64)
    df_expect.index.names = ["date"]
    assert_frame_equal(df, df_expect)


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20150629", "2015-06-30"), ("2015/06/29", "30Jun2015"),
     (pd.datetime(2015, 6, 29), pd.Timestamp(2015, 6, 30))])
@ifbbg
def test_bdh_one_ticker_two_field_longdata(con, start_date, end_date):
    cols = ['PX_LAST', 'VOLUME']
    df = con.bdh('SPY US Equity', cols, start_date, end_date,
                 longdata=True)
    idx = pd.Index(["date", "ticker", "field", "value"])
    data = [["2015-06-29", "2015-06-29", "2015-06-30", "2015-06-30"],
            ["SPY US Equity", "SPY US Equity", "SPY US Equity", "SPY US Equity"],  # NOQA
            ["PX_LAST", "VOLUME", "PX_LAST", "VOLUME"],
            [205.42, 202621332, 205.85, 182925106]]
    df_expect = pd.DataFrame(data=data, index=idx).transpose()
    df_expect.loc[:, "date"] = pd.to_datetime(df_expect.loc[:, "date"])
    df_expect.loc[:, "value"] = np.float64(df_expect.loc[:, "value"])
    assert_frame_equal(df, df_expect)


@ifbbg
def test_bdh_value_errors(con):
    bad_col = "not_a_fld"
    with pytest.raises(ValueError):
        con.bdh("SPY US Equity", bad_col, "20150630", "20150630")

    bad_ticker = "not_a_ticker"
    with pytest.raises(ValueError):
        con.bdh(bad_ticker, "PX_LAST", "20150630", "20150630")


@ifbbg
def test_bdib(con):
    # BBG has limited history for the IntradayBarRequest service so request
    # recent data
    prev_busday = pd.Timestamp(
        pd.np.busday_offset(pd.Timestamp.today().date(), -1)
    )
    ts1 = prev_busday.strftime("%Y-%m-%d") + "T10:00:00"
    ts2 = prev_busday.strftime("%Y-%m-%d") + "T10:20:01"
    df = con.bdib('SPY US Equity', ts1, ts2, event_type="BID", interval=10)

    ts2e = prev_busday.strftime("%Y-%m-%d") + "T10:20:00"
    idx_exp = pd.date_range(ts1, ts2e, periods=3, name="time")
    col_exp = pd.Index(["open", "high", "low", "close", "volume", "numEvents"])

    assert_index_equal(df.index, idx_exp)
    assert_index_equal(df.columns, col_exp)


# REF TESTS
@ifbbg
def test_ref_one_ticker_one_field(con):
    df = con.ref('AUD Curncy', 'NAME')
    df_expect = pd.DataFrame(
        columns=["ticker", "field", "value"],
        data=[["AUD Curncy", "NAME", "Australian Dollar Spot"]]
    )
    assert_frame_equal(df, df_expect)


@pytest.mark.parametrize(
    "date",
    ["20161010", "2016-10-10", "2016/10/10", "10Oct2016",
     pd.datetime(2016, 10, 10), pd.Timestamp(2016, 10, 10,)])
@ifbbg
def test_ref_one_ticker_one_field_override(con, date):
    df = con.ref('AUD Curncy', 'SETTLE_DT',
                 [("REFERENCE_DATE", date)])
    df_expect = pd.DataFrame(
        columns=["ticker", "field", "value"],
        data=[["AUD Curncy", "SETTLE_DT",
              pd.datetime(2016, 10, 12).date()]]
    )
    assert_frame_equal(df, df_expect)


@ifbbg
def test_ref_invalid_field(con):
    with pytest.raises(ValueError):
        con.ref("EI862261 Corp", "not_a_field")


@ifbbg
def test_ref_not_applicable_field(con):
    # test both cases described in
    # https://github.com/matthewgilbert/pdblp/issues/6
    df = con.ref("BCOM Index", ["INDX_GWEIGHT"])
    df_expect = pd.DataFrame(
        [["BCOM Index", "INDX_GWEIGHT", np.NaN]],
        columns=['ticker', 'field', 'value']
    )
    assert_frame_equal(df, df_expect)

    df = con.ref("BCOM Index", ["INDX_MWEIGHT_PX2"])
    df_expect = pd.DataFrame(
        [["BCOM Index", "INDX_MWEIGHT_PX2", np.NaN]],
        columns=['ticker', 'field', 'value']
    )
    assert_frame_equal(df, df_expect)


@ifbbg
def test_ref_invalid_security(con):
    with pytest.raises(ValueError):
        con.ref("NOT_A_TICKER", "MATURITY")


@ifbbg
def test_ref_applicable_with_not_applicable_field(con):
    df = con.ref("BVIS0587 Index", ["MATURITY", "NAME"])
    df_exp = pd.DataFrame(
        [["BVIS0587 Index", "MATURITY", np.NaN],
         ["BVIS0587 Index", "NAME", "CAD Canada Govt BVAL Curve"]],
        columns=["ticker", "field", "value"])
    assert_frame_equal(df, df_exp)


@ifbbg
def test_ref_mixed_data_error(con):
    # calling ref which returns singleton and array data throws error
    with pytest.raises(ValueError):
        con.ref('CL1 Comdty', 'FUT_CHAIN')


# BULKREF TESTS
@pytest.mark.parametrize(
    "date",
    ["20150530", "2015-05-30", "2015/05/30", "30May2015",
     pd.datetime(2015, 5, 30), pd.Timestamp(2015, 5, 30)])
@ifbbg
def test_bulkref_one_ticker_one_field(con, data_path, date):
    df = con.bulkref('BCOM Index', 'INDX_MWEIGHT',
                     ovrds=[("END_DATE_OVERRIDE", date)])
    df_expected = pd.read_csv(
        os.path.join(data_path, "bulkref_20150530.csv")
    )
    pivot_and_assert(df, df_expected)


@pytest.mark.parametrize(
    "date",
    ["20150530", "2015-05-30", "2015/05/30", "30May2015",
     pd.datetime(2015, 5, 30), pd.Timestamp(2015, 5, 30)])
@ifbbg
def test_bulkref_two_ticker_one_field(con, data_path, date):
    df = con.bulkref(['BCOM Index', 'OEX Index'], 'INDX_MWEIGHT',
                     ovrds=[("END_DATE_OVERRIDE", date)])
    df_expected = pd.read_csv(
        os.path.join(data_path, "bulkref_two_fields_20150530.csv")
    )
    pivot_and_assert(df, df_expected)


@ifbbg
def test_bulkref_singleton_error(con):
    # calling bulkref which returns singleton throws error
    with pytest.raises(ValueError):
        con.bulkref('CL1 Comdty', 'FUT_CUR_GEN_TICKER')


@pytest.mark.parametrize("start_date,end_date", [("19860101", "1987-01-01"),
                         ("1986/01/01", "01Jan87"),
                         (pd.datetime(1986, 1, 1), pd.Timestamp(1987, 1, 1))])
@ifbbg
def test_bulkref_null_scalar_sub_element(con, start_date, end_date):
    # related to https://github.com/matthewgilbert/pdblp/issues/32#issuecomment-385555289  # NOQA
    # smoke test to check parse correctly
    ovrds = [("DVD_START_DT", start_date), ("DVD_END_DT", end_date)]
    con.bulkref("101 HK EQUITY", "DVD_HIST", ovrds=ovrds)


@ifbbg
def test_bulkref_empty_field(con):
    df = con.bulkref(["88428LAA0 Corp"], ["INDEX_LIST"])
    df_exp = pd.DataFrame(
        [["88428LAA0 Corp", "INDEX_LIST", np.NaN, np.NaN, np.NaN]],
        columns=["ticker", "field", "name", "value", "position"]
    )
    assert_frame_equal(df, df_exp)


@ifbbg
def test_bulkref_empty_with_nonempty_field_smoketest(con):
    con.bulkref(['88428LAA0 Corp'], ['INDEX_LIST', 'USE_OF_PROCEEDS'])


@ifbbg
def test_bulkref_not_applicable_field(con):
    df = con.bulkref("CL1 Comdty", ["FUT_DLVRBLE_BNDS_ISINS"])
    df_exp = pd.DataFrame(
        [["CL1 Comdty", "FUT_DLVRBLE_BNDS_ISINS", np.NaN, np.NaN, np.NaN]],
        columns=["ticker", "field", "name", "value", "position"]
    )
    assert_frame_equal(df, df_exp)


@ifbbg
def test_bulkref_not_applicable_with_applicable_field_smoketest(con):
    con.bulkref('CL1 Comdty', ['OPT_CHAIN', 'FUT_DLVRBLE_BNDS_ISINS'])


# REF_HIST TESTS
@pytest.mark.parametrize("dates", [("20160104", "2016-01-05"),
                         ("2016/01/04", "05Jan2016"),
                         (pd.datetime(2016, 1, 4), pd.Timestamp(2016, 1, 5))])
@ifbbg
def test_hist_ref_one_ticker_one_field_numeric(con, dates):
    df = con.ref_hist("AUD1M CMPN Curncy", "DAYS_TO_MTY", dates)
    df_expect = pd.DataFrame(
        {"date": ["20160104", "20160105"],
         "ticker": ["AUD1M CMPN Curncy", "AUD1M CMPN Curncy"],
         "field": ["DAYS_TO_MTY", "DAYS_TO_MTY"],
         "value": [33, 32]}
    )
    assert_frame_equal(df, df_expect)


@pytest.mark.parametrize("dates", [("20160104", "2016-01-05"),
                         ("2016/01/04", "05Jan2016"),
                         (pd.datetime(2016, 1, 4), pd.Timestamp(2016, 1, 5))])
@ifbbg
def test_hist_ref_one_ticker_one_field_non_numeric(con, dates):
    df = con.ref_hist("AUD1M CMPN Curncy", "SETTLE_DT", dates)
    df_expect = pd.DataFrame(
        {"date": ["20160104", "20160105"],
         "ticker": ["AUD1M CMPN Curncy", "AUD1M CMPN Curncy"],
         "field": ["SETTLE_DT", "SETTLE_DT"],
         "value": 2 * [pd.datetime(2016, 2, 8).date()]}
    )
    assert_frame_equal(df, df_expect)


# BULKREF_HIST TESTS
@pytest.mark.parametrize(
    "dates",
    [("20150530", "2016-05-30"), ("2015/05/30", "30May16"),
     (pd.datetime(2015, 5, 30), pd.Timestamp(2016, 5, 30))])
@ifbbg
def test_bulkref_hist_one_field(con, data_path, dates):
    df = con.bulkref_hist('BCOM Index', 'INDX_MWEIGHT', dates=dates,
                          date_field='END_DATE_OVERRIDE')
    df_expected = pd.read_csv(
        os.path.join(data_path, "bulkref_20150530_20160530.csv")
    )
    pivot_and_assert(df, df_expected, with_date=True)


@pytest.mark.parametrize(
    "dates", [("20160625",), ("2016-06-25",), ("2016/06/25",), ("25Jun2016",),
              (pd.datetime(2016, 6, 25),), (pd.Timestamp(2016, 6, 25),)])
@ifbbg
def test_bulkhist_ref_with_alternative_reference_field(con, dates):
    # smoke test to  check that the response was sent off and correctly
    # received
    con.bulkref_hist("BVIS0587 Index", "CURVE_TENOR_RATES", dates,
                     date_field="CURVE_DATE")


@pytest.mark.parametrize(
    "start_date,end_date",
    [("20150629", "2015-06-30"), ("2015/06/29", "30Jun2015"),
     (pd.datetime(2015, 6, 29), pd.Timestamp(2015, 6, 30))])
@ifbbg
def test_context_manager(port, host, timeout, start_date, end_date):
    with pdblp.bopen(host=host, port=port, timeout=timeout) as bb:
        df = bb.bdh('SPY US Equity', 'PX_LAST', start_date, end_date)
    midx = pd.MultiIndex(levels=[["SPY US Equity"], ["PX_LAST"]],
                         labels=[[0], [0]], names=["ticker", "field"])
    df_expect = pd.DataFrame(
        index=pd.date_range("2015-06-29", "2015-06-30"),
        columns=midx,
        data=[205.42, 205.85]
    )
    df_expect.index.names = ["date"]
    assert_frame_equal(df, df_expect)


@ifbbg
def test_context_manager_passed_session(port, host):
    sopts = blpapi.SessionOptions()
    sopts.setServerHost(host)
    sopts.setServerPort(port)
    session = blpapi.Session(sopts)
    session.start()
    session.nextEvent(1000)
    session.nextEvent(1000)
    with pdblp.bopen(session=session) as bb:  # NOQA
        pass


@ifbbg
def test_multi_start(port, host, timeout):
    con = pdblp.BCon(host=host, port=port, timeout=timeout)
    con.start()
    con.start()


@ifbbg
def test_non_empty_session_queue(port, host):
    sopts = blpapi.SessionOptions()
    sopts.setServerHost(host)
    sopts.setServerPort(port)
    session = blpapi.Session(sopts)
    session.start()
    with pytest.raises(ValueError):
        pdblp.BCon(session=session)


@ifbbg
def test_bsrch(con):
    df = con.bsrch("COMDTY:VESSEL").head()
    df_expect = pd.DataFrame(["IMO1000019 Index", "LADY K II",
                              "IMO1000021 Index", "MONTKAJ",
                              "IMO1000033 Index"])
    assert_frame_equal(df, df_expect)


def test_connection_error(port):
    con = pdblp.BCon(port=port + 1)
    with pytest.raises(ConnectionError):
        con.start()


pdblp/utils.py
import blpapi


def custom_req(session, request):
    """
    Utility for sending a predefined request and printing response as well
    as storing messages in a list, useful for testing

    Parameters
    ----------
    session: blpapi.session.Session
    request: blpapi.request.Request
        Request to be sent

    Returns
    -------
        List of all messages received
    """
    # flush event queue in case previous call errored out
    while(session.tryNextEvent()):
        pass

    print("Sending Request:\n %s" % request)
    session.sendRequest(request)
    messages = []
    # Process received events
    while(True):
        # We provide timeout to give the chance for Ctrl+C handling:
        ev = session.nextEvent(500)
        for msg in ev:
            print("Message Received:\n %s" % msg)
            messages.append(msg)
        if ev.eventType() == blpapi.Event.RESPONSE:
            # Response completely received, so we could exit
            break
    return messages


pdblp/parser.py
"""
This module provides functionality for parsing the string representations of
Bloomberg Open API Requests and Responses into equivalent JSON or dictionary
representations
"""

# A huge thank you to Paul McGuire who provided invaluable help working out
# the grammar for this. Discussion available at
# https://stackoverflow.com/questions/44144055/parsing-json-like-format-with-pyparsing/44172752#44172752
import json
import pyparsing as pp


def _parse(mystr):

    LBRACE, RBRACE, EQUAL = map(pp.Suppress, "{}=")
    field = pp.Word(pp.printables + ' ', excludeChars='[]=')
    field.addParseAction(pp.tokenMap(str.rstrip))
    string = pp.dblQuotedString().setParseAction(pp.removeQuotes)
    number = pp.pyparsing_common.number()
    date_expr = pp.Regex(r'\d\d\d\d-\d\d-\d\d')
    time_expr = pp.Regex(r'\d\d:\d\d:\d\d\.\d\d\d')
    nan = pp.Keyword('nan')
    scalar_value = (string | date_expr | time_expr | number | nan)

    list_marker = pp.Suppress("[]")
    value_list = pp.Forward()
    jobject = pp.Forward()

    memberDef1 = pp.Group(field + EQUAL + scalar_value)
    memberDef2 = pp.Group(field + EQUAL + jobject)
    memberDef3 = pp.Group(field + list_marker + EQUAL + LBRACE + value_list +
                          RBRACE)
    memberDef = memberDef1 | memberDef2 | memberDef3

    value_list <<= (pp.delimitedList(scalar_value, ",") |
                    pp.ZeroOrMore(pp.Group(pp.Dict(memberDef2))))
    value_list.setParseAction(lambda t: [pp.ParseResults(t[:])])

    members = pp.OneOrMore(memberDef)
    jobject <<= pp.Dict(LBRACE + pp.ZeroOrMore(memberDef) + RBRACE)
    # force empty jobject to be a dict
    jobject.setParseAction(lambda t: t or {})

    parser = members
    parser = pp.OneOrMore(pp.Group(pp.Dict(memberDef)))

    return parser.parseString(mystr)


def to_dict_list(mystr):
    """
    Translate a string representation of a Bloomberg Open API Request/Response
    into a list of dictionaries.return

    Parameters
    ----------
    mystr: str
        A string representation of one or more blpapi.request.Request or
        blp.message.Message, these should be '\\n' seperated
    """
    res = _parse(mystr)
    dicts = []
    for res_dict in res:
        dicts.append(res_dict.asDict())
    return dicts


def to_json(mystr):
    """
    Translate a string representation of a Bloomberg Open API Request/Response
    into a JSON string

    Parameters
    ----------
    mystr: str
        A string representation of one or more blpapi.request.Request or
        blp.message.Message, these should be '\\n' seperated
    """
    dicts = to_dict_list(mystr)
    json.dumps(dicts, indent=2)


examples/example_calls.py
import pdblp

# instantiate and start a connection, set debug=True for printing messages
con = pdblp.BCon(debug=True)
con.start()

# simple bdh call
con.bdh('SPY US Equity', 'PX_LAST', '20150629', '20150630')

# two fields returns MultiIndex
con.bdh('SPY US Equity', ['PX_LAST', 'VOLUME'], '20150629', '20150630')

# bdh with override
con.bdh('MPMIEZMA Index', 'PX_LAST', '20150101', '20150830')
con.bdh('MPMIEZMA Index', 'PX_LAST', '20150101', '20150830',
        ovrds=[('RELEASE_STAGE_OVERRIDE', 'P')])

# reference data request
con.ref('AUDUSD Curncy', 'SETTLE_DT')
con.ref(['NZDUSD Curncy', 'AUDUSD Curncy'], 'SETTLE_DT')
con.ref('AUDUSD Curncy', ['SETTLE_DT', 'DAYS_TO_MTY'])
con.ref(['NZDUSD Curncy', 'AUDUSD Curncy'], ['SETTLE_DT', 'DAYS_TO_MTY'])
con.ref('AUDUSD Curncy', 'SETTLE_DT', [('REFERENCE_DATE', '20150715')])
con.ref(['NZDUSD Curncy', 'AUDUSD Curncy'], ['SETTLE_DT', 'DAYS_TO_MTY'],
        [('REFERENCE_DATE', '20150715')])
# mutli reference data request
con.ref('W 1 Comdty', 'FUT_CHAIN')
con.ref('W 1 Comdty', 'FUT_CHAIN', [('INCLUDE_EXPIRED_CONTRACTS', 'Y')])

# custom request
request = con.refDataService.createRequest("ReferenceDataRequest")
request.append('securities', 'AUD1M Curncy')
request.append('fields', 'DAYS_TO_MTY')
con.custom_req(request)

# add field override
overrides = request.getElement('overrides')
override1 = overrides.appendElement()
override1.setElement("fieldId", 'REFERENCE_DATE')
override1.setElement('value', '20150629')
con.custom_req(request)

# historical reference data request
con.ref_hist('AUD1M Curncy', 'DAYS_TO_MTY', '20150625', '20150629')
con.ref_hist(['AUD1M Curncy', 'NZD1M Curncy'], 'DAYS_TO_MTY', '20150625',
             '20150629')
con.ref_hist('AUD1M Curncy', ['DAYS_TO_MTY', 'SETTLE_DT'], '20150625',
             '20150629')
con.ref_hist(['AUD1M Curncy', 'NZD1M Curncy'], ['DAYS_TO_MTY', 'SETTLE_DT'],
             '20150625', '20150629')
